### Analyte Classification using SVM
#### Edgar Acuna 
#### July 2020

In [25]:
import warnings
warnings.filterwarnings("ignore")
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn import svm
from sklearn.metrics import classification_report, confusion_matrix, f1_score, precision_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt
%matplotlib inline

#### Loading the data

In [26]:
df1=pd.read_csv("c://onr2020/NRLset1_part1.csv",header=None)
df2=pd.read_csv("c://onr2020/NRLset1_part2.csv",header=None)
df3=pd.read_csv("c://onr2020/NRLset1_part3.csv",header=None)
df4=pd.read_csv("c://onr2020/NRLset1_part4.csv",header=None)
df5=pd.read_csv("c://onr2020/NRLset1_part5.csv",header=None)
df6=pd.read_csv("c://onr2020/NRLset1_part6.csv",header=None)
df7=pd.read_csv("c://onr2020/NRLset1_part7.csv",header=None)
df8=pd.read_csv("c://onr2020/NRLset1_part8.csv",header=None)
y=pd.read_csv("c://onr2020/labels.csv",header=None)
ys=pd.read_csv("c://onr2020/substrateIDs.csv",header=None)

In [27]:
dfset1=pd.concat([df1,df2,df3,df4,df5,df6,df7,df8],ignore_index=True)
dfset1.head()

0         1         2         3         4         5         6     \
0  0.041418  0.041621  0.042198  0.042688  0.042924  0.042274  0.042542   
1  0.694710  0.695840  0.697180  0.706530  0.703970  0.709130  0.711480   
2  0.048978  0.048432  0.047685  0.047086  0.046811  0.046752  0.046624   
3  0.039762  0.039495  0.038982  0.038339  0.037769  0.037301  0.036799   
4  0.022387  0.022508  0.022091  0.023054  0.023010  0.022740  0.023889   

       7         8         9     ...      1691      1692      1693      1694  \
0  0.042663  0.042715  0.042664  ...  0.053219  0.053131  0.053091  0.053140   
1  0.711330  0.712370  0.712980  ...  0.754260  0.751040  0.753630  0.754490   
2  0.046443  0.046241  0.045999  ...  0.260320  0.259590  0.258910  0.258210   
3  0.036316  0.035921  0.035612  ...  0.257600  0.256890  0.256160  0.255370   
4  0.023936  0.023464  0.024810  ...  0.025036  0.024307  0.024636  0.022298   

       1695      1696      1697      1698      1699      1700  
0  0.053249  0.053325  0.053364  0.053356  0.053500  0.055986  
1  0.752980  0.755760  0.752910  0.755320  0.753030  0.753230  
2  0.257440  0.256580  0.255700  0.254800  0.254160  0.257110  
3  0.254540  0.253720  0.252880  0.251970  0.251070  0.250190  
4  0.023536  0.025714  0.025306  0.025062  0.023609  0.023901  

[5 rows x 1701 columns]

### SVM to the whole dataset

In [28]:
#print(y.value_counts())
X=dfset1
#lb_make = LabelEncoder()
#dfsub1["Analyte"] = lb_make.fit_transform(dfsub1["Analyte"])
#y2=dfsub1['Analyte']
#print(y2.value_counts())
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,random_state=0)
scaler = StandardScaler()
scaler.fit(X_train)
X_train= scaler.transform(X_train)
X_test = scaler.transform(X_test)
clf=svm.LinearSVC()
clf.fit(X_train, y_train) 
predictions = clf.predict(X_test)
print("F1-score is=",f1_score(y_test,predictions, average="weighted"))
print("precision=",precision_score(y_test,predictions,average="weighted"))
print("The accuracy is=",clf.score(X_test,y_test))
print(classification_report(y_test, predictions))

F1-score is= 0.9307544680292262
precision= 0.9397430788406145
The accuracy is= 0.9272222222222222
              precision    recall  f1-score   support

           1       1.00      0.97      0.99       108
           2       0.86      0.99      0.92        96
           3       0.98      0.91      0.94        98
           4       0.91      0.90      0.91        94
           5       0.94      0.98      0.96        92
           6       1.00      0.97      0.98        96
           7       0.99      0.90      0.94        91
           8       0.98      0.97      0.97        87
           9       0.47      0.95      0.63        83
          10       0.82      0.99      0.90        70
          11       0.97      0.95      0.96        80
          12       0.91      0.91      0.91        88
          13       0.92      0.87      0.89        91
          14       0.89      0.92      0.91        78
          15       0.95      0.91      0.93        78
          16       0.96      0.86    

### Analytes with lowest F1-score: 13 14 18 32 34 36 37 38

In [29]:
#### SVM in  each substrate

In [30]:
dfset2=dfset1.copy()
dfset2['Analyte']=y
dfset2['substrate']=ys
dfsub1=dfset2[dfset2['substrate']==1]
dfsub2=dfset2[dfset2['substrate']==2]
dfsub3=dfset2[dfset2['substrate']==3]
dfsub4=dfset2[dfset2['substrate']==4]
dfsub5=dfset2[dfset2['substrate']==5]
dfsub6=dfset2[dfset2['substrate']==6]
dfsub7=dfset2[dfset2['substrate']==7]
dfsub8=dfset2[dfset2['substrate']==8]
dfsub9=dfset2[dfset2['substrate']==9]
dfsubset0=pd.concat([dfsub3,dfsub4,dfsub9],ignore_index=True)

In [31]:
dfsub61 = dfsub6.drop(dfsub6[(dfsub6['Analyte']==13) | (dfsub6['Analyte']==24) | (dfsub6['Analyte']==34)].index)
dfsub41 = dfsub4.drop(dfsub4[(dfsub4['Analyte']==13) | (dfsub4['Analyte']==24) | (dfsub4['Analyte']==34)].index)
dfsub31 = dfsub3.drop(dfsub3[(dfsub3['Analyte']==13) | (dfsub3['Analyte']==24) | (dfsub3['Analyte']==34)].index)
dfsub11 = dfsub1.drop(dfsub1[(dfsub1['Analyte']==13) | (dfsub1['Analyte']==24) | (dfsub1['Analyte']==34)].index)
dfsub21 = dfsub2.drop(dfsub2[(dfsub2['Analyte']==13) | (dfsub2['Analyte']==24) | (dfsub2['Analyte']==34)].index)
dfsub51 = dfsub5.drop(dfsub5[(dfsub5['Analyte']==13) | (dfsub5['Analyte']==24) | (dfsub5['Analyte']==34)].index)
dfsub71 = dfsub7.drop(dfsub7[(dfsub7['Analyte']==13) | (dfsub7['Analyte']==24) | (dfsub7['Analyte']==34)].index)
dfsub81 = dfsub8.drop(dfsub8[(dfsub8['Analyte']==13) | (dfsub8['Analyte']==24) | (dfsub8['Analyte']==34)].index)
dfsub91 = dfsub9.drop(dfsub9[(dfsub9['Analyte']==13) | (dfsub9['Analyte']==24) | (dfsub9['Analyte']==34)].index)
dfsub81.head()

0         1         2         3         4         5         6  \
17  0.044380  0.044443  0.044488  0.044612  0.044522  0.044302  0.044212   
18  0.042934  0.043964  0.044740  0.046504  0.045221  0.045903  0.046998   
22  0.026433  0.025775  0.027481  0.041315  0.041874  0.033917  0.038958   
25  0.020407  0.039471  0.017211  0.029796  0.052852  0.038253  0.036894   
39  0.042248  0.042862  0.039195  0.042300  0.040911  0.038766  0.042281   

           7         8         9  ...     1693     1694     1695     1696  \
17  0.044010  0.043464  0.042900  ...  0.57174  0.57167  0.57151  0.57145   
18  0.045203  0.046316  0.046820  ...  0.56994  0.56986  0.56923  0.56947   
22  0.045950  0.036473  0.031280  ...  0.56776  0.58419  0.56650  0.57524   
25  0.021150  0.023028  0.029952  ...  0.56682  0.56426  0.58340  0.57052   
39  0.047375  0.036080  0.041750  ...  0.56847  0.57228  0.56452  0.57093   

       1697     1698     1699     1700  Analyte  substrate  
17  0.57140  0.57140  0.57148  0.57214       21          8  
18  0.56902  0.56943  0.56837  0.57062        9          8  
22  0.58549  0.59316  0.56882  0.55885       23          8  
25  0.56990  0.57565  0.57394  0.57924       28          8  
39  0.57087  0.56932  0.57237  0.57107        5          8  

[5 rows x 1703 columns]

In [32]:
from sklearn import svm
y=dfsub1['Analyte']
#print(y.value_counts())
X=dfsub1.iloc[:,0:1701]
#lb_make = LabelEncoder()
#dfsub1["Analyte"] = lb_make.fit_transform(dfsub1["Analyte"])
#y2=dfsub1['Analyte']
#print(y2.value_counts())
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,random_state=0)
scaler = StandardScaler()
scaler.fit(X_train)
X_train= scaler.transform(X_train)
X_test = scaler.transform(X_test)
clf=svm.LinearSVC()
clf.fit(X_train, y_train) 
predictions = clf.predict(X_test)
print("F1-score is=",f1_score(y_test,predictions, average="weighted"))
print("precision=",precision_score(y_test,predictions,average="weighted"))
print("The accuracy is=",clf.score(X_test,y_test))
print(classification_report(y_test, predictions))

F1-score is= 0.9011366110330371
precision= 0.9168269300144299
The accuracy is= 0.8975
              precision    recall  f1-score   support

           1       0.89      0.89      0.89         9
           2       1.00      1.00      1.00        14
           3       1.00      0.89      0.94         9
           4       1.00      0.91      0.95        11
           5       1.00      0.93      0.97        15
           6       1.00      0.85      0.92        13
           7       1.00      0.92      0.96        12
           8       1.00      0.92      0.96        13
           9       1.00      1.00      1.00         8
          10       1.00      0.90      0.95        10
          11       1.00      1.00      1.00         3
          12       0.90      1.00      0.95         9
          13       0.50      0.50      0.50         8
          14       1.00      1.00      1.00         9
          15       0.88      1.00      0.93         7
          16       1.00      0.80      0.89      

### Binary classification

In [33]:
#Only the first twenty analytes are analyzed
labels=list(range(1,41))

In [34]:
for j in labels:
    yclass=dfsub2['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub2.iloc[:,0:1701],yclass,test_size=0.2,random_state=0)
    scaler = StandardScaler()
    scaler.fit(X_train)
    X_train= scaler.transform(X_train)
    X_test = scaler.transform(X_test)
    clf=svm.SVC()
    clf.fit(X_train, yclass_train) 
    predictions = clf.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, clf.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=0.990, F1-score=0.778, Precision=1.000
Metrics for predicting analyte 2 : accuracy=0.983, F1-score=0.533, Precision=1.000
Metrics for predicting analyte 3 : accuracy=0.993, F1-score=0.667, Precision=1.000
Metrics for predicting analyte 4 : accuracy=0.993, F1-score=0.824, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.993, F1-score=0.880, Precision=1.000
Metrics for predicting analyte 6 : accuracy=0.998, F1-score=0.800, Precision=1.000
Metrics for predicting analyte 7 : accuracy=0.983, F1-score=0.588, Precision=1.000
Metrics for predicting analyte 8 : accuracy=0.988, F1-score=0.545, Precision=1.000
Metrics for predicting analyte 9 : accuracy=0.980, F1-score=0.333, Precision=1.000
Metrics for predicting analyte 10 : accuracy=0.983, F1-score=0.588, Precision=1.000
Metrics for predicting analyte 11 : accuracy=0.988, F1-score=0.444, Precision=1.000
Metrics for predicting analyte 12 : accuracy=0.975, F1-score=0.000, Precision=0.000
M

In [35]:
for j in labels:
    yclass=dfsub3['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub3.iloc[:,0:1701],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 2 : accuracy=0.995, F1-score=0.857, Precision=0.750
Metrics for predicting analyte 3 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 4 : accuracy=0.995, F1-score=0.857, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.985, F1-score=0.842, Precision=0.800
Metrics for predicting analyte 6 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 7 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 8 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 9 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 12 : accuracy=0.985, F1-score=0.769, Precision=0.833
M

In [36]:
for j in labels:
    yclass=dfsub6['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub6.iloc[:,0:1701],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=0.995, F1-score=0.909, Precision=1.000
Metrics for predicting analyte 2 : accuracy=0.995, F1-score=0.889, Precision=1.000
Metrics for predicting analyte 3 : accuracy=0.995, F1-score=0.800, Precision=1.000
Metrics for predicting analyte 4 : accuracy=0.995, F1-score=0.667, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.995, F1-score=0.857, Precision=1.000
Metrics for predicting analyte 6 : accuracy=0.995, F1-score=0.800, Precision=1.000
Metrics for predicting analyte 7 : accuracy=0.995, F1-score=0.800, Precision=1.000
Metrics for predicting analyte 8 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 9 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 12 : accuracy=1.000, F1-score=1.000, Precision=1.000
M

In [37]:
labels1=[1,2,3,4,5,6,7,8,9,10,11,12,14,15,16,17,18,19,20,21,22,23,25,26,27,28,29,30,31,32,33,35,36,37,38,39,40]
for j in labels1:
    yclass=dfsub71['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub71.iloc[:,0:1701],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 2 : accuracy=0.995, F1-score=0.667, Precision=1.000
Metrics for predicting analyte 3 : accuracy=0.989, F1-score=0.667, Precision=0.500
Metrics for predicting analyte 4 : accuracy=0.984, F1-score=0.769, Precision=1.000
Metrics for predicting analyte 5 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 6 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 7 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 8 : accuracy=0.989, F1-score=0.857, Precision=1.000
Metrics for predicting analyte 9 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=0.995, F1-score=0.933, Precision=1.000
Metrics for predicting analyte 12 : accuracy=0.984, F1-score=0.667, Precision=0.750
M

In [38]:
for j in labels1:
    yclass=dfsub81['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub81.iloc[:,0:1701],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=0.995, F1-score=0.857, Precision=1.000
Metrics for predicting analyte 2 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 3 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 4 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.995, F1-score=0.667, Precision=1.000
Metrics for predicting analyte 6 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 7 : accuracy=0.995, F1-score=0.800, Precision=0.667
Metrics for predicting analyte 8 : accuracy=0.995, F1-score=0.889, Precision=1.000
Metrics for predicting analyte 9 : accuracy=0.989, F1-score=0.875, Precision=0.778
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=0.995, F1-score=0.947, Precision=0.900
Metrics for predicting analyte 12 : accuracy=0.989, F1-score=0.833, Precision=1.000
M

In [24]:
for j in labels1:
    yclass=dfsub91['Analyte'].copy()
    yclass[yclass!=j]=0 
    yclass[yclass==j]=1
    X_train, X_test, yclass_train, yclass_test = train_test_split(dfsub91.iloc[:,0:1701],yclass,test_size=0.1,random_state=0)
    neigh = KNeighborsClassifier(n_neighbors=3)
    neigh.fit(X_train, yclass_train) 
    predictions = neigh.predict(X_test)
    print("Metrics for predicting analyte %d : accuracy=%.3f, F1-score=%.3f, Precision=%.3f" %(j, neigh.score(X_test,yclass_test), f1_score(yclass_test,predictions),precision_score(yclass_test,predictions)))

Metrics for predicting analyte 1 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 2 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 3 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 4 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 5 : accuracy=0.995, F1-score=0.889, Precision=1.000
Metrics for predicting analyte 6 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 7 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 8 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 9 : accuracy=0.995, F1-score=0.933, Precision=1.000
Metrics for predicting analyte 10 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 11 : accuracy=1.000, F1-score=1.000, Precision=1.000
Metrics for predicting analyte 12 : accuracy=1.000, F1-score=1.000, Precision=1.000
M

#### The F1-score is above .90 for six  analytes and,  between .80-.90 for 22 analytes .  The analytes 13(0.652),18(0.618),24(0.602) and 34(0.601) have the lowest F1-score. Also, they have the lowest auc.

#### The prediction results improve. Now, the F1-score is above .90 for twenty-eight analytes and, between .80-.90 for nine analytes . The analytes 13(0.757), 24(0.602) and 34(0.760) have the lowest F1-score. Also, they have the lowest auc.